src/run_graph.rs

```diff
@@ -264,27 +271,45 @@ pub fn run_ex<DB: WriteResults + Sync>(
                 loop {
-                    let option_task = graph.lock().unwrap().next_task(ex, db);
-                    if let Some((id, task)) = option_task {
-                        info!("running task: {:?}", task);
-                        if let Err(e) = task.run(config, ex, db) {
-                            error!("task failed, marking childs as failed too: {:?}", task);
-                            util::report_error(&e);
-
-                            let result = if config.is_broken(&task.krate) {
-                                TestResult::BuildFail
+                    let walk_result = graph.lock().unwrap().next_task(ex, db);
+                    match walk_result {
+                        WalkResult::Task(id, task) => {
+                            info!("running task: {:?}", task);
+                            if let Err(e) = task.run(config, ex, db) {
+                                error!("task failed, marking childs as failed too: {:?}", task);
+                                util::report_error(&e);
+
+                                let result = if config.is_broken(&task.krate) {
+                                    TestResult::BuildFail
+                                } else {
+                                    TestResult::Error
+                                };
+                                graph
+                                    .lock()
+                                    .unwrap()
+                                    .mark_as_failed(id, ex, db, &e, result)?;
                             } else {
-                                TestResult::Error
-                            };
-                            graph
-                                .lock()
-                                .unwrap()
-                                .mark_as_failed(id, ex, db, &e, result)?;
-                        } else {
-                            graph.lock().unwrap().mark_as_completed(id);
+                                graph.lock().unwrap().mark_as_completed(id);
+                            }
+
+                            // Unpark all the threads
+                            let mut parked = parked_threads.lock().unwrap();
+                            for thread in parked.drain(..) {
+                                thread.unpark();
+                            }
+                        }
+                        WalkResult::Blocked => {
+                            // Wait until another thread finished before looking for tasks again
+                            // If the thread spuriously wake up (parking does not guarantee no
+                            // spurious wakeups) it's not a big deal, it will just get parked again
+                            {
+                                let mut parked_threads = parked_threads.lock().unwrap();
+                                parked_threads.push(thread::current());
```

Comment:
Don't we get a bunch of duplicates in here because it's a Vec? Maybe should be a set instead.

Reply:
Oh, yeah, we can use a map there. Fixed!